pkg/cli/cli.go

```diff
@@ -1232,6 +1244,126 @@ func runConnectionWebrtcAgentCommand(flags *ConnectFlags, c *command, args []str
 	return nil
 }
 
+func runConnectionWebSocketAgentCommand(flags *ConnectFlags, c *command, args []string, opts *subCommandOpts) error {
+	// Change Web URL to WebSocket URL
+	// e.g. https://127.0.0.1:1443/ -> wss://127.0.0.1:1443
+	wsRoot := flags.ServiceURL
+	wsRoot = strings.ReplaceAll(wsRoot, "https://", "wss://")
+	wsRoot = strings.ReplaceAll(wsRoot, "http://", "ws://")
+	wsRoot, _ = strings.CutSuffix(wsRoot, "/")
+
+	// Connect to ADB proxy WebSocket
+	//   wss://127.0.0.1:1443/devices/cvd-1/adb
+	// Since the operator is self-signed, skip verifying cert
+	dialer := websocket.Dialer {
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+	device := args[0]
+	wsUrl := fmt.Sprintf("%s/devices/%s/adb", wsRoot, device)
+	wsConn, _, err := dialer.Dial(wsUrl, nil)
+	if err != nil {
+		log.Fatal("Failed to connect ", wsUrl, ": ", err)
+		return err
+	}
+	defer wsConn.Close()
+
+	// Open local TCP port that ADB server connects to and make ADB server connect
+	// to the port
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		c.PrintErrf("Failed to listen ADB socket: %v", err)
+		return err
+	}
+	adbPort := l.Addr().(*net.TCPAddr).Port
+	go func() {
+		if err := opts.ADBServerProxy.Connect(adbPort); err != nil {
+			c.PrintErrf("Failed to connect ADB to device: %v\n", err)
+		}
+	}()
+	tcpConn, err := l.Accept()
+	if err != nil {
+		c.PrintErrf("Failed to accept ADB socket: %v", err)
+		return err
+	}
+	defer tcpConn.Close()
+
+	// Send connect result to the parent
+	result := ConnStatus {
+		ADB: ForwarderState {
+			Port: adbPort,
+			State: StateAsStr(FwdConnected),
+		},
+	}
+	output, err := json.Marshal(result)
+	if err != nil {
+		return err
+	} else {
+		c.Println(string(output))
+	}
+
+	// Close all standard IOs to make agent as daemon
+	if cin, ok := c.InOrStdin().(io.Closer); ok {
+		cin.Close()
+	}
+	if cout, ok := c.OutOrStdout().(io.Closer); ok {
+		cout.Close()
+	}
+	if cerr, ok := c.ErrOrStderr().(io.Closer); ok {
+		cerr.Close()
+	}
+
+	// Redirect WebSocket to ADB TCP port
+	wsWrapper := &wsIoWrapper{
+		wsConn: wsConn,
+		pos: 0,
+		buf: nil,
+	}
```

Comment:
`defer wsWrapper.Close()`?

Reply:
Added `defer wsWrapper.Close()` inside go func() { ... } to clean-up in the go routine side.